KAFKA-7632: Support Compression Level

dongjinleekr opened the pull request:
Since I reworked [KIP-390](https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Support+Compression+Level) from scratch, here I open a new PR.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dongjinleekr committed:
Add CompressionConfig and separate compression configs per codec: NoneConfig, GzipConfig, SnappyConfig, LZ4Config, ZstdConfig.

- Now, each compression codec can have independent config entries.
- Make MemoryRecordsBuilder, KafkaRaftClient, SnapshotWriter to use CompressionConfig instead of CompressionType.
- Make LogAppendInfo to use (CompressionType, CompressionConfig) for compression source and target.
- Add ProducerConfig#getCompressionConfig: parses compression configs and returns per-codec CompressionConfig from given configurations.
- Add LogConfig#getCompressionConfig: parses compression configs and returns per-codec CompressionConfig from given configurations.
- Separate CompressedRecordBatchValidationBenchmark into [Gzip, Snappy, LZ4, Zstd]CompressedRecordBatchValidationBenchmark

dongjinleekr commented:
@ijuma @guozhangwang Could you have a look when you are free? I refined the original work with a benchmark with a real-world dataset. As you can see in the updated KIP, this option can improve the producer's performance significantly. :pray:

tombentley commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -183,8 +183,8 @@ public double compressionRatio() {
 
-    public CompressionType compressionType() {
-        return compressionType;
+    public CompressionConfig compressionConfing() {
```
typo, it should be `compressionConfig()`

tombentley commented on clients/src/main/java/org/apache/kafka/common/record/CompressionConfig.java:
```diff
@@ -0,0 +1,70 @@
+        this.type = type;
+
+        if (level != null && !type.isValidLevel(level.intValue())) {
+            throw new IllegalArgumentException("Illegal level " + level + " for compression codec " + type.name);
```
Should the message include the valid levels?

dongjinleekr replied:
Let me have a look.

tombentley commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -933,6 +934,9 @@ object KafkaConfig {
   "'producer' which means retain the original compression codec set by the producer."
+  val CompressionLevelDoc = "The compression level for all data generated by the producer. The default level and valid value is up to " +
+    "compression.type. (<code>none</code>, <code>snappy</code>: not available. <code>gzip</code>: 1~9. <code>lz4</code>: 1~17. " +
+    "<code>zstd</code>: -131072~22."
```
Since the docs use `[min,...,max]` syntax for things validated by `ConfigDef.Range` I think we should use a similar syntax here for consistency (even though it's not validated by `Range`).

kkonstantine commented:
This PR implements [KIP-390](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=97550583) but is still open and has several conflicts. Unless I hear from you by EOD Thursday, July 8 (pacific time), I'll remove it from the KIP list for 3.0

dongjinleekr commented:
@kkonstantine Here it is - I rebased it onto the latest trunk. Could anyone review this PR? :pray:

kkonstantine commented:
@tombentley do you think you'll be able to finish the review soon? If we had an approval I'd be willing to make an exception and consider this feature for inclusion to AK 3.0.

dongjinleekr committed:
Make GzipConfig to support compression level

dongjinleekr committed:
Make LZ4Config to support compression level

1. Remove unused constructor in KafkaLZ4BlockOutputStream. (and now it is used for the compression level instead.)
2. Add KafkaLZ4BlockOutputStream.[MIN_COMPRESSION_LEVEL, MAX_COMPRESSION_LEVEL, DEFAULT_COMPRESSION_LEVEL] (from LZ4Factory#highCompressor and LZ4Constants.[DEFAULT_COMPRESSION_LEVEL, MAX_COMPRESSION_LEVEL]

dongjinleekr committed:
Make ZstdConfig to support compression level

dongjinleekr committed:
Add compression.level configuration

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -312,6 +320,7 @@ public class ProducerConfig extends AbstractConfig {
                                         Importance.LOW,
                                         ACKS_DOC)
                                 .define(COMPRESSION_TYPE_CONFIG, Type.STRING, "none", Importance.HIGH, COMPRESSION_TYPE_DOC)
+                                .define(COMPRESSION_LEVEL_CONFIG, Type.STRING, "", Importance.MEDIUM, COMPRESSION_LEVEL_DOC)
```
The KIP says the default value is null.

dongjinleekr replied:
Could you kindly review the discussion [here](https://github.com/apache/kafka/pull/10826#discussion_r917108950)?

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -523,6 +532,45 @@ boolean idempotenceEnabled() {
         return userConfiguredTransactions || idempotenceEnabled;
     }
 
+    public CompressionConfig getCompressionConfig(CompressionType compressionType) {
```
This is a public class. We probably don't want to expose this as a public method since we don't expose other config values.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/CompressionConfig.java:
```diff
@@ -0,0 +1,101 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.record;
```
Hmm, it seems this class should be in common.compress package instead of here?

dongjinleekr replied:
At first I tried so, but it has the following problems:

1. Consistency with `org.apache.kafka.common.record.CompressionType`.
2. `CompressionConfig` uses `org.apache.kafka.common.record.(MemoryRecordsBuilder|RecordBatch)` but only the imports from `org.apache.kafka.common.record` to `org.apache.kafka.common.compress` are allowed.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/CompressionConfig.java:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Wrap buffer with an InputStream that will decompress data with this CompressionType.
+     *
+     * @param decompressionBufferSupplier The supplier of ByteBuffer(s) used for decompression if supported.
```
Could we add the rest of the params to javadoc?

junrao commented on clients/src/main/java/org/apache/kafka/common/compress/GzipOutputStream.java:
```diff
@@ -0,0 +1,48 @@
+     * @throws IOException If an I/O error has occurred.
+     */
+    public GzipOutputStream(OutputStream out, int size, int level) throws IOException {
+        super(out, size);
```
Could we just directly use the constructor with level and avoid this wrapper class?

dongjinleekr replied:
I tried so, but the `Deflater` instance is defined as `protected` in [DeflaterOutputStream](https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/java/util/zip/DeflaterOutputStream.java). It's why.

junrao commented on clients/src/main/java/org/apache/kafka/common/record/CompressionConfig.java:
```diff
@@ -0,0 +1,101 @@
+    // Instantiating this class directly is disallowed; use builder methods.
+    protected CompressionConfig() {}
+
+    public abstract CompressionType getType();
```
We typically don't use getter. So, this can just by type().

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -31,6 +31,9 @@
 import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.common.record.CompressionConfig;
+import org.apache.kafka.common.record.MemoryRecords;
+import org.apache.kafka.common.record.MemoryRecordsBuilder;
```
Do we need to move those imports?

dongjinleekr replied:
Fixed. For automatic formatting + import ordering, please have a look on [here](https://github.com/apache/kafka/pull/10428).

junrao commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -104,7 +104,7 @@ public final class RecordAccumulator {
     public RecordAccumulator(LogContext logContext,
                              int batchSize,
-                             CompressionType compression,
+                             CompressionConfig compression,
```
Could we adjust the corresponding javadoc?

mimaison replied:
Could we rename this argument to `compressionConfig`?

junrao commented on clients/src/main/java/org/apache/kafka/common/compress/KafkaLZ4BlockOutputStream.java:
```diff
@@ -122,7 +119,11 @@ public KafkaLZ4BlockOutputStream(OutputStream out) throws IOException {
     }
 
     public KafkaLZ4BlockOutputStream(OutputStream out, boolean useBrokenHC) throws IOException {
```
This is an existing issue. But this constructor and the one above seem unused and could be removed.

dongjinleekr replied:
Sure. Removed.

mimaison commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -188,6 +190,12 @@ public class ProducerConfig extends AbstractConfig {
+    /** <code>compression.level</code> */
+    public static final String COMPRESSION_LEVEL_CONFIG = "compression.level";
+    private static final String COMPRESSION_LEVEL_DOC = "The compression level for all data generated by the producer. The default level and valid value is up to "
+        + "compression.type. (<code>none</code>, <code>snappy</code>: not available. <code>gzip</code>: 1~9. <code>lz4</code>: 1~17. "
```
Should we include the defaults too?

ijuma replied:
I think the reason why @dongjinleekr didn't include those is that they may change at the compression library level.

mimaison commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -523,6 +532,45 @@ boolean idempotenceEnabled() {
     }
 
+    public CompressionConfig getCompressionConfig(CompressionType compressionType) {
+        if (getString(ProducerConfig.COMPRESSION_LEVEL_CONFIG).isEmpty()) {
```
Since only 3 cases use the compression level, what about moving the if/else inside each case block?

```suggestion
        public CompressionConfig getCompressionConfig(CompressionType compressionType) {
        String compressionLevel = getString(ProducerConfig.COMPRESSION_LEVEL_CONFIG);
        switch (compressionType) {
            case NONE:
                return CompressionConfig.NONE;
            case GZIP:
                if (compressionLevel.isEmpty())
                    return CompressionConfig.gzip().build();
                else
                    return CompressionConfig.gzip().setLevel(Integer.parseInt(compressionLevel)).build();
            case SNAPPY:
                return CompressionConfig.snappy().build();
            case LZ4:
                if (compressionLevel.isEmpty())
                    return CompressionConfig.lz4().build();
                else
                    return CompressionConfig.lz4().setLevel(Integer.parseInt(compressionLevel)).build();
            case ZSTD:
                if (compressionLevel.isEmpty())
                    return CompressionConfig.zstd().build();
                else
                    return CompressionConfig.zstd().setLevel(Integer.parseInt(compressionLevel)).build();
            default:
                throw new IllegalArgumentException("Unknown compression type: " + compressionType.name);
        }
    }
```

ijuma replied:
Why don't we push some of these things to the enum itself? `CompressionType.config(String level)` or something like that.

dongjinleekr replied:
The reason why I did not choose a simple approach like `CompressionType.config(String level)` (as @ijuma pointed out) here is simple: **To make room for supporting per-codec configuration in the future, like [KIP-780](https://cwiki.apache.org/confluence/display/KAFKA/KIP-780%3A+Support+fine-grained+compression+options).**

The reason why I chose to use `compression.level` without default value (as @mimaison pointed out) is also simple: at the point of when [KIP-390](https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Support+Compression+Level) is passed, the proposal included global configuration option (i.e., `compression.level`). But, as I work on [KIP-780](https://cwiki.apache.org/confluence/display/KAFKA/KIP-780%3A+Support+fine-grained+compression+options) and running an [in-house](https://en.wikipedia.org/wiki/Naver_Corporation) preview, I found that per-codec configuration (i.e., `compression.{gzip|lz4|zstd}.level`) is much more preferable for the following reasons:

### 1. Easy to switch the compression type

For example, `{compression.type=zstd,compression.level=10}` works but `{compression.type=gzip,compression.level=10}` does not, since gzip only supports the level of 1 to 9. In other words, `compression.type` is so error-prone when the user tries to switch the codec. (@junrao indicated a similar problem: [#1](https://github.com/apache/kafka/pull/10826#discussion_r850624372) [#2](https://github.com/apache/kafka/pull/10826#discussion_r850624372))

If we limit the scope of per-codec options like `compression.gzip.level`, we can easily switch the whole options by changing `compression.type` only. Moreover, this approach is consistent with [additional compression options](https://cwiki.apache.org/confluence/display/KAFKA/KIP-780%3A+Support+fine-grained+compression+options) like `compression.gzip.buffer`.

### 2. Can provide a default level.

As @ijuma pointed out, we can not provide a default compression level with `compression.level`; but with `compression.{gzip|lz4|zstd}.level`, **we can**.

For these reasons, I run the overall benchmarks in [here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-780%3A+Support+fine-grained+compression+options) with per-config options (see 'Final Notes' section.) If the reviewers prefer this scheme, I will happily change it.

mimaison replied:
Thanks @dongjinleekr for the explanations. My suggested change was logically identical, it's just a bit shorter.

mimaison commented on clients/src/main/java/org/apache/kafka/common/compress/KafkaLZ4BlockOutputStream.java:
```diff
@@ -41,6 +41,10 @@ public final class KafkaLZ4BlockOutputStream extends OutputStream {
 
     public static final String CLOSED_STREAM = "The stream is already closed";
 
+    public static final int MIN_COMPRESSION_LEVEL = 1;
```
Should we rename this class to `LZ4OutputStream` to be consistent with `GzipOutputStream`?

ijuma replied:
The kafka naming convention would be `Lz4OutputStream`.

mimaison commented on clients/src/main/java/org/apache/kafka/common/compress/ZstdConfig.java:
```diff
@@ -32,21 +34,39 @@
+     *
+     * For details, please refer the official zstd manual: http://facebook.github.io/zstd/zstd_manual.html
+     */
+    static final int MAX_COMPRESSION_LEVEL = 22;
```
Should we use `Zstd.maxCompressionLevel()` instead of hard coding the value?

mimaison commented on clients/src/main/java/org/apache/kafka/common/compress/KafkaLZ4BlockOutputStream.java:
```diff
@@ -41,6 +41,10 @@ public final class KafkaLZ4BlockOutputStream extends OutputStream {
     public static final String CLOSED_STREAM = "The stream is already closed";
 
+    public static final int MIN_COMPRESSION_LEVEL = 1;
+    public static final int MAX_COMPRESSION_LEVEL = 17;
```
Should we use constants from `LZ4Constants` instead of hardcoding values?

dongjinleekr replied:
I tried to, but `LZ4Constants` is not accessable from outside of the package. ([#](https://github.com/lz4/lz4-java/blob/master/src/java/net/jpountz/lz4/LZ4Constants.java))

mimaison reviewed (commented):
Thanks for the PR!

I've only looked at the changes in clients so far. It looks good overall, I've left a few questions. Can you also rebase on trunk?

ijuma commented:
@mimaison fyi, I think we need more votes in the KIP thread.

mimaison commented:
As far as I can tell this is implementing KIP-390 and it received 3 binding votes: https://lists.apache.org/thread/wnd6ky3kjzv9gkzc39qy5gg2pp20ovs1

ijuma commented:
@mimaison yeah, I just edited my message. :)

ijuma commented on clients/src/main/java/org/apache/kafka/common/compress/GzipConfig.java:
```diff
@@ -0,0 +1,87 @@
+        this.level = level;
+    }
+
+    public CompressionType getType() {
```
We avoid `get` prefixes in kafka.

ijuma commented on clients/src/main/java/org/apache/kafka/common/compress/GzipOutputStream.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * An extension of {@link GZIPOutputStream}, with compression level functionality.
+ */
+public class GzipOutputStream extends GZIPOutputStream {
```
Do we really need to create a subclass? It doesn't seem to add much value.

ijuma replied:
I guess there's no way to access `def` without subclassing?

dongjinleekr replied:
Oh yes, `Deflater def` is defined as `protected` in [DeflaterOutputStream](https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/java/util/zip/DeflaterOutputStream.java). It's why.

junrao commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -59,7 +59,7 @@ class GroupMetadataManager(brokerId: Int,
                            metrics: Metrics) extends Logging with KafkaMetricsGroup {
 
-  private val compressionType: CompressionType = CompressionType.forId(config.offsetsTopicCompressionCodec.codec)
+  private val compressionConfig: CompressionConfig = CompressionConfig.of(CompressionType.forId(config.offsetsTopicCompressionCodec.codec)).build
```
The KIP doesn't mention this. For offset topic, should we stick with the default compression level or introduce a separate config for compression level?

dongjinleekr replied:
I think introducing a separate config for compression level is somewhat overkill; so let's use the default compression level for the offset topic. (current - I will update the KIP accordingly later.)

junrao commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -177,6 +179,44 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
+    if (this.compressionLevel.isEmpty) {
+      compressionType match {
+        case CompressionType.NONE => CompressionConfig.NONE
+        case CompressionType.GZIP => CompressionConfig.gzip.build
```
For methods with side effect like build, we want to include (). So, build() instead of build.

junrao commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -96,6 +97,7 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
   val uncleanLeaderElectionEnable = getBoolean(LogConfig.UncleanLeaderElectionEnableProp)
   val minInSyncReplicas = getInt(LogConfig.MinInSyncReplicasProp)
   val compressionType = getString(LogConfig.CompressionTypeProp).toLowerCase(Locale.ROOT)
+  val compressionLevel = getString(LogConfig.CompressionLevelProp)
```
With topic level overriding, things can be a bit tricky. Suppose that a user sets up a broker level compression type and compression level, but only overrides the compression type for a topic. With this logic, the topic will pick up the broker level compression level, which could be intended for a different compression type.

dongjinleekr replied:
ditto: [here](https://github.com/apache/kafka/pull/10826#discussion_r917108950)

junrao commented on core/src/main/scala/kafka/log/UnifiedLog.scala:
```diff
@@ -1131,19 +1130,19 @@ class UnifiedLog(@volatile var logStartOffset: Long,
 
     // Apply broker-side compression if any
-    val targetCodec = BrokerCompressionCodec.getTargetCompressionCodec(config.compressionType, sourceCodec)
+    val targetConfig = config.compressionConfig(sourceType)
```
Hmm, I am not sure about this change. The old logic handles the case when the topic level compression type is producer. In that case, the producer codec will be used. The new code doesn't seem to handle that.

dongjinleekr replied:
Oh yes, `LogConfing#compressionConfig` now does the thing:

```
def compressionConfig(producerCompressionType: CompressionType): CompressionConfig = {
    compressionType match {
        case ProducerCompressionCodec.name => getCompressionConfig(producerCompressionType)
        case "uncompressed" => CompressionConfig.NONE
        case _ => getCompressionConfig(CompressionType.forName(compressionType))
    }
}
```

But I will double check the test cases.

junrao commented on core/src/main/scala/kafka/log/LogConfig.scala:
```diff
@@ -177,6 +179,44 @@ case class LogConfig(props: java.util.Map[_, _], overriddenConfigs: Set[String]
+  /**
+   * Returns per-codec [[CompressionConfig]] for given `compressionType`.
+   */
+  private def getCompressionConfig(compressionType: CompressionType): CompressionConfig = {
```
Another thing is that if we set a topic or broker level compression type to be the same as the producer, but with a different compression level, whether we should trigger broker side recompression. Currently, we don't do recompression. Should we do it? If so, is it possible to know the producer compression level?

dongjinleekr replied:
No, we can't trigger broker side recompression in these scenario since there is no field for compression level in `RecordBatch`. (Oh, this should be also addressed in the KIP.)

junrao commented on raft/src/main/java/org/apache/kafka/snapshot/SnapshotWriter.java:
```diff
@@ -113,12 +113,11 @@ private void finalizeSnapshotWithFooter() {
      * Create an instance of this class and initialize
      * the underlying snapshot with {@link SnapshotHeaderRecord}
      *
-     * @param snapshot a lambda to create the low level snapshot writer
```
This is an existing issue. Could we add the missing javadoc for supplier?

dongjinleekr replied:
Fixed by rebasing onto the latest trunk.

junrao reviewed (commented):
@dongjinleekr : Thanks for the PR and sorry for the late review. Made a pass of all non-testing files. A few comments below.

dongjinleekr committed:
Rename class name: KafkaLZ4BlockOutputStream → Lz4OutputStream

dongjinleekr committed:
Change method names: getType → type. setLevel → level.

dongjinleekr committed:
Trivial improvements

- Change scope of 'ProducerConfig#getCompressionConfig(CompressionType)' into package.
- Revert import order + Fix parameter name in RecordAccumulator.
- Add Javadoc params to CompressionConfig#wrapForInput.
- Fix Javadoc param in RecordsSnapshotWriter.
- Add '()' to 'CompressionCodec.Builder#build' call in scala.

dongjinleekr committed:
Remove unused constructors in Lz4OutputStream

dongjinleekr commented:
Sorry for being late. Rebased onto the latest trunk, with resolving the issues raised by @mimaison @ijuma @junrao.

mimaison commented on clients/src/main/java/org/apache/kafka/common/config/TopicConfig.java:
```diff
@@ -169,6 +169,14 @@ public class TopicConfig {
+     * <code>compression.level</code>
+     */
+    public static final String COMPRESSION_LEVEL_CONFIG = "compression.level";
+    public static final String COMPRESSION_LEVEL_DOC = "The compression level for all data generated by the producer. The default level and valid value is up to "
```
What about:
```
The compression level to use when compressing records broker side. This is only used if compression.type is set to gzip, lz4 or zstd. The default level and valid value is up to...
```

mimaison commented on clients/src/test/java/org/apache/kafka/clients/producer/internals/RecordAccumulatorTest.java:
```diff
@@ -289,8 +290,7 @@ private void testAppendLargeOldMessageFormat(CompressionType compressionType) th
         int lingerMs = 10;
         RecordAccumulator accum = createTestRecordAccumulator(
-                1024 + DefaultRecordBatch.RECORD_BATCH_OVERHEAD, 10 * 1024, CompressionType.NONE, lingerMs);
-        accum.append(topic, partition1, 0L, key, value, Record.EMPTY_HEADERS, null, maxBlockTimeMs, false, time.milliseconds(), cluster);
```
Don't we need this line for the test to pass?

mimaison commented on clients/src/main/java/org/apache/kafka/common/compress/ZstdConfig.java:
```diff
@@ -70,4 +91,22 @@ public void release(ByteBuffer buffer) {
+        private int level = DEFAULT_COMPRESSION_LEVEL;
+
+        public Builder level(int level) {
+            if (MAX_COMPRESSION_LEVEL < level) {
```
Should we also check the min level?

mimaison commented on clients/src/main/java/org/apache/kafka/common/compress/LZ4Config.java:
```diff
@@ -0,0 +1,79 @@
+
+        public Builder level(int level) {
+            if (level < Lz4OutputStream.MIN_COMPRESSION_LEVEL || Lz4OutputStream.MAX_COMPRESSION_LEVEL < level) {
+                throw new IllegalArgumentException("lz4 doesn't support given compression level: " + level);
```
Can we include the min and max values in the error message?

mimaison commented on clients/src/test/java/org/apache/kafka/common/compress/KafkaLZ4Test.java:
```diff
@@ -113,9 +116,11 @@ public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
-                            for (boolean close : Arrays.asList(false, true))
-                                arguments.add(Arguments.of(new Args(broken, ignore, blockChecksum, close, payload)));
+                        // Available levels: 1, 2, ... 17.
+                        for (int compressionLevel = 1; compressionLevel < 18; ++compressionLevel)
```
Should we use `Lz4OutputStream.MIN_COMPRESSION_LEVEL` and `MAX_COMPRESSION_LEVEL` rather than hardcoding the levels?

mimaison commented on core/src/test/scala/other/kafka/TestLinearWriteSpeed.scala:
```diff
@@ -101,14 +108,21 @@ object TestLinearWriteSpeed {
     val createTime = System.currentTimeMillis
+    val properties = new Properties()
+    properties ++= extraProps
+    properties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9000")
```
Ah right this is only created in order to call `getCompressionConfig`. For a second I was wondering where `localhost:9000` came from.

mimaison reviewed (commented):
Thanks @dongjinleekr for the updates. I've made another pass and left a few minor comments.

clolov commented:
Hello @dongjinleekr! What is the current state of this pull request?

ozturkberkay commented:
Hi, I was wondering if this PR would be prioritized in the foreseeable future given there has been a lot of work done already

junrao commented:
One discussion of the PR is around choosing the default value for `compression.level` (https://github.com/apache/kafka/pull/10826#discussion_r917108950). This is tricky since it's specific to the compression type. I am wondering if it's better to adjust the KIP to support `compression.{gzip|lz4|zstd}.level` instead. This will make it more aligned with the future KIP https://cwiki.apache.org/confluence/display/KAFKA/KIP-780%3A+Support+fine-grained+compression+options#KIP780:Supportfinegrainedcompressionoptions-Zstd.1.

mimaison commented:
Yes it would be good to get this over the line. @dongjinleekr Do you think you'll have time to update it? or would you need some help?

mimaison commented:
I updated the [KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-390%3A+Support+Compression+Level) based on [Jun's suggestion](https://github.com/apache/kafka/pull/10826#issuecomment-1795952612) and sent an update to the VOTE thread.

mimaison commented:
I opened https://github.com/apache/kafka/pull/15516 to implement this KIP.

mimaison commented:
Closing this PR since https://github.com/apache/kafka/pull/15516 has been merged.

dongjinleekr commented:
@junrao @mimaison @clolov @ozturkberkay

I greatly appreciate you finalizing this PR! After changing jobs, I could not get involved with the Kafka community since I had to focus on our in-house fork, aka Navercorp (@naver) Kafka.

Now, I just restarted KIP-780 and am testing it with our in-house tiered storage plugin. As soon as I complete the work, I will re-open the PR. Thank you again, and stay tuned! :bowing_man:
